lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp

```diff
@@ -0,0 +1,294 @@
+  std::lock_guard<std::recursive_mutex> guard(m_mutex);
+  Log *log = GetLog(LLDBLog::Types);
+  auto it = m_info_to_module.find(info_id);
+
+  // We might've registered and loaded the cache for this module already. Return
+  // early in that case.
+  if (it != m_info_to_module.end()) {
+    if (std::get<bool>(it->second)) {
+      LLDB_LOGV(
+          log,
+          "[SwiftMetadataCache] Module {0} already register, returning early.",
+          module->GetFileSpec().GetFilename());
+      return;
+    }
+  }
+
+  m_info_to_module[info_id] = {module, false};
+  loadCacheForModule(module, info_id);
+}
+
+void SwiftMetadataCache::loadCacheForModule(ModuleSP module, uint64_t info_id) {
+  std::lock_guard<std::recursive_mutex> guard(m_mutex);
+
+  Log *log = GetLog(LLDBLog::Types);
+
+  if (!m_data_file_cache)
+    return;
+
+  auto module_name = getTyperefCacheFileNameForModule(module);
+  auto mem_buffer_up = m_data_file_cache->GetCachedData(module_name);
+
+  // Nothing cached.
+  if (!mem_buffer_up) {
+    LLDB_LOGV(log, "[SwiftMetadataCache] No cached file found for module {0}.",
+              module->GetFileSpec().GetFilename());
+    return;
+  }
+
+  // Extractor used to extract the header information (see the .h file for
+  // details on the format).
+  DataExtractor header_extractor(mem_buffer_up->getBufferStart(),
+                                 mem_buffer_up->getBufferSize(),
+                                 module->GetObjectFile()->GetByteOrder(),
+                                 module->GetObjectFile()->GetAddressByteSize());
+
+  lldb::offset_t read_offset = 0;
+
+  std::string UUID = module->GetUUID().GetAsString();
+  std::string cached_UUID = header_extractor.GetCStr(&read_offset);
+  // If no uuid in the file something is wrong with the cache.
+  if (cached_UUID.empty()) {
+    LLDB_LOG(log,
+             "[SwiftMetadataCache] Failed to read cached UUID for module {0}.",
+             module->GetFileSpec().GetFilename());
+    m_data_file_cache->RemoveCacheFile(module_name);
+    return;
+  }
+
+  // If the UUIDs don't match this is most likely a stale cache.
+  if (cached_UUID != UUID) {
+    LLDB_LOGV(log, "[SwiftMetadataCache] Module UUID mismatch for {0}.",
+              module->GetFileSpec().GetFilename());
+    m_data_file_cache->RemoveCacheFile(module_name);
+    return;
+  }
+
+  uint64_t expanded_size = 0;
+  if (!header_extractor.GetU64(&read_offset, &expanded_size, 1)) {
+    LLDB_LOGV(log,
+              "[SwiftMetadataCache] Failed to read decompressed cache size for "
+              "module {0}.",
+              module->GetFileSpec().GetFilename());
+    m_data_file_cache->RemoveCacheFile(module_name);
+    return;
+  }
+
+  const auto *start = (const char *)header_extractor.GetData(&read_offset, 0);
+  // Create a reference to the compressed data.
+  llvm::StringRef string_buffer(start, (uint64_t)mem_buffer_up->getBufferEnd() -
+                                           (uint64_t)start);
```

Comment:
membuffer_up->getBuffer() ?
https://llvm.org/doxygen/classllvm_1_1MemoryBuffer.html#a2ea075bd68f15b57e95f0771b8ba0bca

Reply:
Do you mean `getBufferSize`? If so, we can't do that because we have to subtract the header we just read.

Reply:
ah, I didn't realize that start is not the start of the buffer,